# Manual differentiation

In [1]:
import numpy as np

def f(x1,x2):
    return x1*x2+np.sin(x1)

In [2]:
f(1,2)

2.8414709848078967

In [3]:
def manual_dfdx1(x1,x2):
    return x2+np.cos(x1)

def manual_dfdx2(x1,x2):
    return x1

In [4]:
manual_dfdx1(1,2)

2.5403023058681398

In [5]:
manual_dfdx2(1,2)

1

# Symbolic differentiation

## Symbolic computation of the derivative

In [7]:
from sympy import *

In [8]:
x1 = Symbol('x1')
x2 = Symbol('x2')
g = x1*x2+sin(x1)

In [9]:
type(x1)

sympy.core.symbol.Symbol

In [10]:
type(g)

sympy.core.add.Add

In [11]:
sym_dgdx1 = diff(g,x1)

In [12]:
type(sym_dgdx1)

sympy.core.add.Add

In [13]:
sym_dgdx2 = diff(g,x2)

In [14]:
type(sym_dgdx2)

sympy.core.symbol.Symbol

In [15]:
print("g(x1,x2) = "+str(g))
print("dg/dx1   = "+str(sym_dgdx1))
print("dg/dx2   = "+str(sym_dgdx2))

g(x1,x2) = x1*x2 + sin(x1)
dg/dx1   = x2 + cos(x1)
dg/dx2   = x1


In [16]:
print("dg/dx1 (1.0,2.0) = " + str(sym_dgdx1.evalf(subs={x1: 1.0, x2:2.0})))
print("dg/dx2 (1.0,2.0) = " + str(sym_dgdx2.evalf(subs={x1: 1.0, x2:2.0})) )

dg/dx1 (1.0,2.0) = 2.54030230586814
dg/dx2 (1.0,2.0) = 1.00000000000000


## Simplification of functions is possible as well

In [18]:
x = Symbol('x')
y = Symbol('y')
g = (x+x*y)/x
print("g = "+str(f))
print("simplified = "+str(simplify(g)))

g = <function f at 0x7f4aec6c8a60>
simplified = y + 1


## Limits of symbolic differentiation

In [19]:
import numpy as np

def h(x1,x2):
    result = 0
    if x1<=x2: # x1*x2 + sin(x1)
        result = x1*x2+np.sin(x1)        
    else: # pi + x1*x2 + 5*x1^2
        result = np.pi;
        result += x1*x2;
        for i in range(0,5):
            result += x1**2
    return result

In [20]:
f(1,2)

2.8414709848078967

In [21]:
h(1,2)

2.8414709848078967

In [22]:
f(2,1)

2.909297426825682

In [23]:
h(2,1)

25.141592653589793

# Numerical differentiation

In [24]:
def numerical_dhdx1(x1,x2):
    s = 0.00001
    return (h(x1+s,x2) - h(x1,x2)) / s

In [25]:
numerical_dhdx1(1,2)

2.5402980984967627

In [26]:
manual_dfdx1(1,2)

2.5403023058681398

We can see that the numerical approximation of the derivative of function h with respect to x1 at point (1,2) is not exactly the same what we would get with the actual derived formulas for the derivative.

Perhaps, s was too large!

In [27]:
def numerical_dhdx1(x1,x2,s):
    return (h(x1+s,x2) - h(x1,x2)) / s


s = 1.0
for i in range(1,20):
    s = s/10.0
    print("numerical_dhdx1(1,2,{0:.20f}) = {1:.20f}".
          format(s, numerical_dhdx1(1,2,s))
         )

numerical_dhdx1(1,2,0.10000000000000000555) = 2.49736375253538867014
numerical_dhdx1(1,2,0.01000000000000000021) = 2.53608598101182636242
numerical_dhdx1(1,2,0.00100000000000000002) = 2.53988148035988459128
numerical_dhdx1(1,2,0.00010000000000000000) = 2.54026023141840084918
numerical_dhdx1(1,2,0.00001000000000000000) = 2.54029809849676269096
numerical_dhdx1(1,2,0.00000100000000000000) = 2.54030188484577434949
numerical_dhdx1(1,2,0.00000010000000000000) = 2.54030226187751351219
numerical_dhdx1(1,2,0.00000001000000000000) = 2.54030227964108190619
numerical_dhdx1(1,2,0.00000000100000000000) = 2.54030219082323949209
numerical_dhdx1(1,2,0.00000000010000000000) = 2.54030130264481979196
numerical_dhdx1(1,2,0.00000000001000000000) = 2.54027909818432728883
numerical_dhdx1(1,2,0.00000000000100000000) = 2.54019028034235772040
numerical_dhdx1(1,2,0.00000000000010000000) = 2.53574938824385709424
numerical_dhdx1(1,2,0.00000000000001000000) = 2.53130849614535646808
numerical_dhdx1(1,2,0.000000000000

# Automatic differentiation

## Forward mode autodiff

### Simple example

In [28]:
import numpy as np

def f(x1,x2):
    """
    f(x1,x2)=x1*x2+sin(x1)
    """
    w1 = x1
    dw1 = 1
    
    w2 = x2
    dw2 = 0
    
    w3 = w1*w2
    dw3 = dw1 * w2 + w1 * dw2
    
    w4 = np.sin(w1)
    dw4 = np.cos(w1) * dw1
    
    w5 = w3 + w4
    dw5 = dw3 + dw4
    
    return w5, dw5

val, deriv = f(1,2)
print("Result according to forward mode autodiff")
print("df/dx1 (1,2)=" + str(deriv))

print("Result according to manual differation:")
# Manual derivation: df/dx1 = x2 + cos(x1)
print("df/dx1 (1,2)=" + str(2 + np.cos(1)))

Result according to forward mode autodiff
df/dx1 (1,2)=2.5403023058681398
Result according to manual differation:
df/dx1 (1,2)=2.5403023058681398


In [29]:
import numpy as np

def f(x1,x2):
    """
    f(x1,x2)=x1*x2+sin(x1)
    """
    w1 = x1
    dw1 = 0 # changed this ...
    
    w2 = x2
    dw2 = 1 # ... and that
    
    w3 = w1*w2
    dw3 = dw1 * w2 + w1 * dw2
    
    w4 = np.sin(w1)
    dw4 = np.cos(w1) * dw1
    
    w5 = w3 + w4
    dw5 = dw3 + dw4
    
    return w5, dw5

val, deriv = f(1,2)
print("Result according to forward mode autodiff")
print("df/dx2 (1,2)=" + str(deriv))

print("Result according to manual differation:")
# Manual derivation: df/dx2 = x1
print("df/dx2 (1,2)=" + str(1))

Result according to forward mode autodiff
df/dx2 (1,2)=1.0
Result according to manual differation:
df/dx2 (1,2)=1


### Complex example with loops:

In [30]:
import numpy as np

def h(x1,x2):    
    result = 0
    if x1<=x2:
        # h(x1,x2):=x1*x2 + sin(x1)
        # Manual differentiation:
        # dh/dx1 = x2 + cos(x1)
        # dh/dx2 = x1
        result = x1*x2+np.sin(x1)        
    else: 
        # h(x1,x2):=pi + x1*x2 + 5*x1^2
        # Manual differentiation:
        # dh/dx1 = x2+10*x1
        # dh/dx2 = x1
        result = np.pi;
        result += x1*x2;
        for i in range(0,5):
            result += x1**2
    return result

In [31]:
h(1,2)

2.8414709848078967

In [44]:
1*2+np.sin(1)

2.8414709848078967

In [32]:
h(2,1)

25.141592653589793

In [46]:
np.pi+2*1+5*(2**2)

25.141592653589793

Now we reformulate the computation in function h with the w-notation:

In [33]:
import numpy as np

def h(x1,x2):   
    w1 = x1
    w2 = x2
    
    if x1<=x2:
        # h(x1,x2):=x1*x2 + sin(x1)
        # Manual differentiation:
        # dh/dx1 = x2 + cos(x1)
        # dh/dx2 = x1
        w3 = w1*w2
        w4 = np.sin(w1)
        w5 = w3+w4
        return w5
        
    else: 
        # h(x1,x2):=pi + x1*x2 + 5*x1^2
        # Manual differentiation:
        # dh/dx1 = x2+10*x1
        # dh/dx2 = x1
        w3 = np.pi
        w4 = w1*w2
        w5 = w3+w4
        
        w6 = 0
        for i in range(0,5):
            w7 = w1**2
            w6 = w6+w7
            
        w8 = w5+w6
        return w8

In [34]:
h(1,2)

2.8414709848078967

In [35]:
h(2,1)

25.141592653589793

Now let us augment the code, such that we can compute the partial derivatives with the forward mode:

In [48]:
import numpy as np

def h(x1,x2):   
    w1 = x1
    w2 = x2
    dw1 = 1
    dw2 = 0
    
    if x1<=x2:
        # h(x1,x2):=x1*x2 + sin(x1)
        # Manual differentiation:
        # dh/dx1 = x2 + cos(x1)
        # dh/dx2 = x1
        w3 = w1*w2
        dw3 = dw1*w2 + w1*dw2
        
        w4 = np.sin(w1)
        dw4 = np.cos(w1)*dw1
        
        w5 = w3+w4
        dw5 = dw3+dw4
        
        return w5,dw5
        
    else: 
        # h(x1,x2):=pi + x1*x2 + 5*x1^2
        # Manual differentiation:
        # dh/dx1 = x2+10*x1
        # dh/dx2 = x1
        w3 = np.pi
        dw3 = 0
        
        w4 = w1*w2
        dw4 = dw1*w2 + w1*dw2
        
        w5 = w3+w4
        dw5 = dw3+dw4
        
        w6 = 0
        dw6 = 0
        
        for i in range(0,5):
            w7 = w1**2
            dw7 = 2*w1
            
            w6 = w6+w7
            dw6 = dw6+dw7
            
        w8 = w5+w6
        dw8 = dw5+dw6
        
        return w8, dw8

In [49]:
h(1,2)

(2.8414709848078967, 2.5403023058681398)

In [50]:
# dh/dx1 = x2 + cos(x1)
2+np.cos(1)

2.5403023058681398

In [51]:
h(2,1)

(25.141592653589793, 21)

In [52]:
# dh/dx1 = x2+10*x1
1+10*2

21

## Reverse mode autodiff

### Simple example

Let us take our old function f() as an example:

In [53]:
import numpy as np
def f(x1,x2):
    return x1*x2+np.sin(x1)

In [54]:
f(1,2)

2.8414709848078967

Let us rewrite it with the w-notation:

In [55]:
def f(x1,x2):
    
    # forward step
    w1 = x1
    w2 = x2
    w3 = w1*w2
    w4 = np.sin(w1)
    w5 = w3+w4
    
    return w5

In [56]:
f(1,2)

2.8414709848078967

Now let us add the computation nodes for the dy/dw's:

In [57]:
def f(x1,x2):
    
    # forward step
    w1 = x1
    w2 = x2
    w3 = w1*w2
    w4 = np.sin(w1)
    w5 = w3+w4
    
    # backward step
    _w5 = 1
    _w4 = _w5*1
    _w3 = _w5*1
    _w2 = _w3*w1
    
    _w1a = _w3*w2
    _w1b = _w4*np.cos(w1)
    _w1 = _w1a+_w1b
    
    return w5,_w1,_w2

In [58]:
f(1,2)

(2.8414709848078967, 2.5403023058681398, 1)

Both derivatives df/dx1=2.540302 and df/dx2=1.0 are correct.

### Complex example

In [59]:
def f(x1,x2):
    
    # forward step
    w1 = x1
    w2 = x2
    w3 = w1*w2
    w4 = np.sin(w1)
    w5 = w3*w4  # CHANGED note! now: multiplication!
    
    # backward step
    _w5 = 1
    _w4 = _w5*w3 # CHANGED
    _w3 = _w5*w4 # CHANGED
    _w2 = _w3*w1
    
    _w1a = _w3*w2
    _w1b = _w4*np.cos(w1)
    _w1 = _w1a+_w1b
    
    return w5,_w1,_w2

In [60]:
f(1,2)

(1.682941969615793, 2.7635465813520725, 0.8414709848078965)